test/hotspot/jtreg/compiler/c2/TestUnsignedByteCompare.java

```diff
@@ -23,72 +23,67 @@
 
 /*
  * @test
- * @bug 8204479
- * @summary Bitwise AND on byte value sometimes produces wrong result
+ * @bug 8204479 8253191
  *
- * @run main/othervm -XX:+IgnoreUnrecognizedVMOptions -XX:-TieredCompilation
- *      -XX:-UseOnStackReplacement -XX:-BackgroundCompilation -Xcomp -XX:-Inline
- *      compiler.c2.TestUnsignedByteCompare
+ * @library /test/lib
+ * @modules java.base/jdk.internal.vm.annotation
+ *
+ * @run main/bootclasspath/othervm -XX:+IgnoreUnrecognizedVMOptions -XX:-TieredCompilation compiler.c2.TestUnsignedByteCompare
  */
-
 package compiler.c2;
 
+import java.lang.invoke.*;
+import jdk.internal.vm.annotation.DontInline;
+import jdk.test.lib.Asserts;
+
 public class TestUnsignedByteCompare {
 
-    static int p, n;
+    @DontInline static boolean testByteGT0(byte[] val) { return (val[0] & mask()) >  0; }
+    @DontInline static boolean testByteGE0(byte[] val) { return (val[0] & mask()) >= 0; }
+    @DontInline static boolean testByteEQ0(byte[] val) { return (val[0] & mask()) == 0; }
+    @DontInline static boolean testByteNE0(byte[] val) { return (val[0] & mask()) != 0; }
+    @DontInline static boolean testByteLE0(byte[] val) { return (val[0] & mask()) <= 0; }
+    @DontInline static boolean testByteLT0(byte[] val) { return (val[0] & mask()) <  0; }
 
-    static void report(byte[] ba, int i, boolean failed) {
-        // Enable for debugging:
-        // System.out.println((failed ? "Failed" : "Passed") + " with: " + ba[i] + " at " + i);
+    static void testValue(byte b) {
```

Comment:
Shouldn't you exclude that method from compilation to compare interpreted vs. C2 compiled result?

Reply:
I could do that, but it's not required: `testUB_mem_imm` matches memory operand, but `testValue` does 
`(b & mask())` while `testByte...` do `val[0] & mask()`. So, `testUB_mem_imm` is used only in `testByte...` methods.